text/3668-async-closure.md

```diff
@@ -0,0 +1,541 @@
+
+For example:
+
+```rust
+let s = String::from("hello, world");
+// Implements `async Fn()` along with `FnMut` and `Fn`
+// because it can copy the `&String` that it captures.
+let _ = async || {
+    println!("{s}");
+};
+
+let s = String::from("hello, world");
+// Implements `async Fn()` but not `FnMut` or `Fn` because
+// it moves and owns a value of type `String`, and therefore
+// the future it returns needs to take a pointer to data
+// owned by the closure.
+let _ = async move || {
+    println!("{s}");
+};
+
+let mut s = String::from("hello, world");
+// Implements `async FnMut()` but not `FnMut` or `Fn`
+// because it needs to reborrow a mutable pointer to `s`.
+let _ = async move || {
+    s.push('!');
+};
+```
+
+#### Specifics about the `AsyncFnOnce` implementation, interaction with `move`
+
+If the closure is inferred to be `async Fn` or `async FnMut`, then the compiler will synthesize an `async FnOnce` implementation for the closure which returns a future that doesn't borrow any captured values from the closure, but instead *moves* the captured values into the future.
+
+For example:
+
+```rust
+let s = String::from("hello, world");
+
+let closure = async move || {
+    ready(&s);
+};
+// At this point, `s` is moved out of.  However, the
+// allocation for `s` is still live.  It just lives as a
+// captured field in `closure`.
+
+// Manually call `AsyncFnOnce` -- this isn't stable since
+// `AsyncFnOnce` isn't stable, but it's useful for the demo.
+let fut = AsyncFnOnce::call_once(closure, ());
+// At this point, `closure` is dropped.  However, the
+// allocation for `s` is still live.  It now lives as a
+// captured field in `fut`.
+
+fut.await;
+// After the future is awaited, it's dropped.  At that
+// point, the allocation for `s` is dropped.
+```
+
+### Interaction with return-type notation, naming the future returned by calling
+
+With `async Fn() -> T` trait bounds, we don't know anything about the `Future` returned by calling the async closure other than that it's a `Future` and awaiting that future returns `T`.
+
+This is not always sufficient, for example, if you want to spawn a future onto another thread:
+
+```rust
+async fn foo(x: impl async Fn(&str)) -> Result<()> {
+    tokio::spawn(x("hello, world")).await
+}
+```
+
+```
+error[E0277]: cannot be sent between threads safely
+   --> src/lib.rs
+    |
+    |     tokio::spawn(x("hello, world")).await
+    |     ------------ ^^^^^^^^^^^^^^^^^ cannot be sent between threads safely
+    |     |
+    |     required by a bound introduced by this call
+```
+
+When the RTN RFC is accepted, this RFC specifies that users will be allowed to add RTN-like bounds to type parameters that are also bounded by `async Fn()`, like so:
+
+```rust
+async fn foo(x: F) -> Result<()>
+where
+    F: async Fn(&str) -> Result<()>,
+    // The future from calling `F` is `Send` and `'static`.
+    F(..): Send + 'static,
```

Comment:
the bound is applied on both `CallOnceFuture` + `CallRefFuture<'a>`?

Reply:
Yep. I'll detail that -- this section was light since I uplifted it from the "future" section since the RTN RFC is landing soon.